File: consync/profiles.py
import os
from os import path

from consync.resource import Resource


class Profiles:
    def __init__(self, config):
        self.config = config
        self.basepath = config['common']['basepath']
        self.profilesdir = config['common'].get("profilesdir", os.path.join(self.basepath, "profiles"))
        self.profiles = []
        if 'profiles' in config:
           self.profiles = [profile.strip() for profile in config['profiles']['active'].split(",") if profile.strip()]
        print("active profiles: " + " ".join(self.profiles))

    def read(self, resource):
        for profile in self.profiles:
            filepath = path.join(self.profilesdir, profile, resource.path)
            if path.exists(filepath):
                with open(filepath, "rb") as f:
                    if len(resource.content) > 0:
                        resource.content += b'\n'
                    resource.content += f.read()

    def collect_resources(self, resources):
        existing_pathes = [res.path for res in resources]
        for profile in self.profiles:
            profiledir = os.path.join(self.profilesdir, profile)
            for root, dirs, files in os.walk(profiledir, topdown=True):
                dirs[:] = [d for d in dirs if not d.startswith(".")]
                for file in files:
                    filepath = os.path.join(root, file)
                    key = os.path.relpath(filepath, profiledir)
                    if not key in existing_pathes:
                        resource = Resource(os.path.relpath(filepath, profiledir))
                        resource.sources.append(filepath)
                        resources.append(resource)
                    else:
                        resource = [resource for resource in resources if resource.path == key][0]
                        resource.sources.append(filepath)
        return resources

    def filter_resources(self, resources):
        return resources

    def transform_content(self, resources, resource):
        pass

File: consulsync/profiles.go
package consync

import (
	"github.com/pelletier/go-toml"
	"path/filepath"
	"strings"
	"path"
	"os"
	"io/ioutil"
	"fmt"
)

type ProfilePlugin struct {
	rootdir       string
	configuration *toml.TomlTree
	profiles      []string
}

func (plugin ProfilePlugin) collectResources(resources []Resource) ([]Resource, error) {
	if len(plugin.profiles) == 0 {
		return resources, nil
	}
	result := resources
	resourcesByPath := make(map[string]*Resource)
	for _, resource := range resources {
		resourcesByPath[resource.name] = &resource
	}
	for _, profile := range (plugin.profiles) {
		configdir := path.Join(plugin.rootdir, (plugin.configuration.GetDefault("profiles.dir", "profiles")).(string), profile)

		err := filepath.Walk(configdir, func(path string, file os.FileInfo, err error) error {
			if (!file.IsDir()) {
				relpath, _ := filepath.Rel(configdir, path)
				if (!strings.HasPrefix(relpath, ".")) {
					if _, ok := resourcesByPath[relpath]; ok {
						resourcesByPath[relpath].sources = append(resourcesByPath[relpath].sources, path)
					} else {
						resource := Resource{relpath, path, "", []string{relpath}, 0}
						result = append(result, resource)
					}
				}
			}
			return nil
		})

		if (err != nil) {
			return result, err
		}
	}
	return result, nil
}

func (plugin ProfilePlugin) read(resource *Resource) error {
	for _, profile := range (plugin.profiles) {
		path := path.Join(plugin.rootdir, (plugin.configuration.GetDefault("profiles.dir", "profiles")).(string), profile, resource.name)
		if _, err := os.Stat(path); !os.IsNotExist(err) {
			content, err := ioutil.ReadFile(path)
			if (err != nil) {
				return err
			}
			resource.content = fmt.Sprintf("%s\n%s", resource.content, string(content))
		}
	}
	return nil
}

func (plugin ProfilePlugin) transformContent(resources *[]Resource, resource *Resource) error {
	return nil
}
